wasm-rpc/src/lib.rs

```diff
@@ -304,6 +447,1012 @@ fn build_tree(node: &WitNode, nodes: &[WitNode]) -> Value {
+        let mut possible_mapping_indexed: HashMap<&String, &Option<AnalysedType>> = HashMap::new();
+
+        for (name, optional_type) in types.iter() {
+            possible_mapping_indexed.insert(name, optional_type);
+        }
+
+        let json_obj = input_json
+            .as_object()
+            .ok_or(vec![format!("Input {} is not an object", input_json)])?;
+
+        let (key, json) = if json_obj.is_empty() {
+            Err(vec!["Zero variants in in the input".to_string()])
+        } else {
+            Ok(json_obj.iter().next().unwrap())
+        }?;
+
+        match possible_mapping_indexed.get(key) {
+            Some(Some(tpe)) => TypeAnnotatedValue::from_json_value(json, tpe).map(|result| {
+                TypeAnnotatedValue::Variant {
+                    typ: types.to_vec(),
+                    case_name: key.clone(),
+                    case_value: Some(Box::new(result)),
+                }
+            }),
+            Some(None) if json.is_null() => Ok(TypeAnnotatedValue::Variant {
+                typ: types.to_vec(),
+                case_name: key.clone(),
+                case_value: None,
+            }),
+            Some(None) => Err(vec![format!("Unit variant {key} has non-null JSON value")]),
+            None => Err(vec![format!("Unknown key {key} in the variant")]),
+        }
+    }
+
+    fn get_handle(
+        value: &JsonValue,
+        id: AnalysedResourceId,
+        resource_mode: AnalysedResourceMode,
+    ) -> Result<TypeAnnotatedValue, Vec<String>> {
+        match value.as_str() {
+            Some(str) => {
+                // not assuming much about the url format, just checking it ends with a /<resource-id-u64>
+                let parts: Vec<&str> = str.split('/').collect();
+                if parts.len() >= 2 {
+                    match u64::from_str(parts[parts.len() - 1]) {
+                        Ok(resource_id) => {
+                            let uri = parts[0..(parts.len() - 1)].join("/");
+                            Ok(TypeAnnotatedValue::Handle { id, resource_mode, uri: Uri { value: uri }, resource_id })
+                        }
+                        Err(err) => {
+                            Err(vec![format!("Failed to parse resource-id section of the handle value: {}", err)])
+                        }
+                    }
+                } else {
+                    Err(vec![format!(
+                        "Expected function parameter type is Handle, represented by a worker-url/resource-id string. But found {}",
+                        str
+                    )])
+                }
+            }
+            None => Err(vec![format!(
+                "Expected function parameter type is Handle, represented by a worker-url/resource-id string. But found {}",
+                type_description(value)
+            )]),
+        }
+    }
+}
+
+#[cfg(feature = "typeinfo")]
+fn type_description(value: &JsonValue) -> &'static str {
```

Comment:
Having all these top level functions also make me like the recommendation I wrote above, to move all these stuff into a private module and just export the public api with `pub use` from `lib.rs`